### ISSUE TITLE
Add table-watch functionality

### ISSUE THREAD

**Issue by otoolep** [rqlite/rqlite#127]
An interesting idea for a new feature would be to allow clients to watch for changes to the nodes, over a persistent TCP connection, or webhook type feature. As changes are made, the clients would be notified.

**Comment by otoolep** [comment]
SQLite has pre-update and update hooks (seemingly). This could be interesting functionality too, that might make this easier.

**Comment by mainframe** [comment]
+1 this would be extremely useful feature.

**Comment by varung** [comment]
is it possible to enable the client to listen to the RAFT log as a witness, but not actually participate in the quorum? I know with native RAFT this is possible. Might give you the functionality you want for free.

**Comment by otoolep** [comment]
Yes, I have considered having the system stream the Raft commands out over, say, a TCP connection. But I am thinking something more focused would be more useful.

**Comment by zhangyongding** [comment]
+1 this would be extremely useful feature.

**Comment by austinpena-taloflow** [comment]
+1 would enjoy this feature

**Comment by otoolep** [comment]
What would folks actually do with it? What are the applications? If I understood, it might help be better implement it.

**Comment by austinpena-taloflow** [comment]
I'm building a distributed health check application. Failed checks head to the database. It would be nice to be alerted when a new entry is added.

**Comment by otoolep** [comment]
That helps. How would you expect to configure what to be told about, or would you just expect all INSERT and UPDATEs to be streamed out over the network?

**Comment by austinpena-taloflow** [comment]
Others might be better at giving recommendations, but I'll give it a shot:

I think ideally, a solution would use a (pub sub)[https://github.com/ThreeDotsLabs/watermill-sql/] solution. Where all changes are published to a new in memory table and is read. I think the table should be configurable to be in memory, even if the application is on disk.

**Comment by gabriel-samfira** [comment]
Adding my 2 cents here. This would be useful for distributed systems that need to notify clients connected via websocket or grpc of changes that may have happened to an object in the database, without polling.

If you have 3 api servers and a request to change an object in the db came in through server 1, clients connected to servers 2 and 3 would also need to be notified of the change. Something like the postgres listen/notify system would help remove the need to either implement some kind of sync channel in the app itself, or rely on another external component that offers amqp/pubsub.

It's ok if by the time the server reacts to the notify event, the data is already stale. Especially if for each change a notify is sent to each listening party. Eventually events are consumed and clients are in sync. Heck, clients could just fetch a fresh copy of the data from the db if a notify is sent. The point is to know that a change has happened to a table. Maybe include the primary key of the row that has changed. Or even better would be to allow some stored procedure to marshal the data that should be sent as a notification. That way a user could include just the columns they care about.

A size limit for the body of the notification would be great. That way you don't end up sending large binary blobs.

**Comment by otoolep** [comment]
>If you have 3 api servers and a request to change an object in the db came in through server 1, clients connected to servers 2 and 3 

Your mental model of rqlite doesn't seem to be quite right. Clients are not "connected" to servers. They can make requests against any server, but there is no concept of servers having "currently connected clients".

If I build this (and I would like to), it would be basically be the same as CockroachDB -- you register webhooks (or something similar) to be called when changes happen.

**Comment by gabriel-samfira** [comment]
I did a poor job of explaining what I had in mind. Let me try that one more time. But before I do, I must mention that I did not mean to pile more work on you :grinning:. It just feels like a potentially useful feature that is present in some systems like [PostgreSQL](https://www.postgresql.org/docs/current/sql-notify.html) and [etcd](https://etcd.io/docs/v3.5/tutorials/how-to-watch-keys/) and from which rqlite might benefit. RQLite is already very useful, so if this is not a good fit, feel free to ignore me :smile:.

---

The model I had in mind is something like this:

![sketch](https://github.com/rqlite/rqlite/assets/903160/c7d8dad8-4486-4542-9fa5-7ee704acece8)


The clients I was referring to are the ones that may have a websocket or streaming gRPC session with the application that uses rqlite as a database. In this scenario, if `Client 1`, `Client 3` and `Client 5` are interested in a particular entity that is stored in the database, and each of them are connected to a different `API/Controller` node (greatly simplified sketch), when `Client 1` updates the entity of interest, it would be great to be able to notify `Client 3` and `Client 5`. If `Client 3` and `Client 5` are connected to different `API/Controller` nodes, those `API/Controller` nodes would have to either poll the database or use some other means of determining when a change has happened.

Traditionally, this can be solved by simply having a shared message bus (amqp, pubsub, etc) between the `API/controller` nodes, and once the transaction returns success, the node that executed the update can also publish a fan-out message on that shared bus. At that point, the rest of the nodes can service any clients that are interested in that update.

Having a "watch" functionality in rqlite (perhaps via a long lived websocket/gRPC connection - given rqlite has a http api), even in a basic form (pk x in table y was updated), would allow for applications to not have to implement a shared message bus via some other component.

Thanks for the amazing project!

**Comment by otoolep** [comment]
OK, this sounds more like what I would like to do. Thanks for the extra detail.

**Comment by otoolep** [comment]
https://philipotoole.com/rqlite-9-0-real-time-change-data-capture-for-distributed-sqlite/
